### PR TITLE
[WIP] Update: ignore padding support for TransfoXL training when n_clusters==0

### PR DESCRIPTION
**Pull request by StefanHeng**
# What does this PR do?

This PR solves [an issue](https://github.com/huggingface/transformers/issues/17446) I raised about TransformerXL. As @sgugger mentioned in [another issue](https://github.com/huggingface/transformers/issues/19914) I raised, he [says](https://github.com/huggingface/transformers/issues/19914#issuecomment-1293656206)

> I don't think TransformerXL supports FP16 as this is an old model with very specific code for the softmax layer. This won't be an issue we will fix ourselves given that Transformer-XL is not very used anymore, but if someone wants to make a PR, we'll review!

I'm using TransformerXL in a [research project](https://github.com/StefanHeng/Symbolic-Music-Generation) and disabling the adaptive softmax is an option I would like to explore.  So here I am. 

In the `n_clusters==0` branch, the current TransformerXL implementation does not work with padding (-100), it beaks at `.gather(1, labels)`. 

This PR solves that bug. 

I tested with my research data and confirmed my implementation is working. It's able to overfit the training data to up to 99% next-token prediction on multiple hyper-parameter setups, for #samples from 8 to 48, batch size from 48 to 64, epochs from 128 to 512. 

Fixes # (issue)
[Issue 19914](https://github.com/huggingface/transformers/issues/19914)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger 
@patrickvonplaten
@thomwolf

### CODE REVIEW

File: src/transformers/models/transfo_xl/modeling_transfo_xl_utilities.py [hunk 1]
```diff
@@ -109,7 +110,13 @@ def forward(self, hidden, labels=None, keep_order=False):
         if self.n_clusters == 0:
             logit = self._compute_logit(hidden, self.out_layers[0].weight, self.out_layers[0].bias, self.out_projs[0])
             if labels is not None:
-                out = -nn.functional.log_softmax(logit, dim=-1).gather(1, labels.unsqueeze(1)).squeeze(1)
+                if padding_index is not None:
+                    mask = labels != padding_index
```
Comment (patrickvonplaten) [inline src/transformers/models/transfo_xl/modeling_transfo_xl_utilities.py:114]: We should probably better check here if:
```python
mask = lables != -100
```
to be more in line with other models no? 
What do you think @sgugger ?

Comment (sgugger) [reply src/transformers/models/transfo_xl/modeling_transfo_xl_utilities.py:114]: Yes, that would be more consistent with how other models/PyTorch loss functions ignore some labels. Though it doesn't work the examples where we encourage `labels=inputs["input_ids"]`

**Review by sgugger — COMMENTED** [review]
Thanks a lot for working on this! Can you make sure to run `make style` on your branch? You will also need to rebase on main to fix all the TensorFlow-related tests (they broke with last TensorFlow release).

**Comment by StefanHeng** [comment]
Thank you for your comments! 

> Thanks a lot for working on this! Can you make sure to run `make style` on your branch? You will also need to rebase on main to fix all the TensorFlow-related tests (they broke with last TensorFlow release).

Sure will do.

**Comment by StefanHeng** [comment]
>  Though it doesn't work the examples where we encourage labels=inputs["input_ids"]

@sgugger Can you clarify? Not sure what do you mean

**Comment by sgugger** [comment]
Please use -100 for padding in the labels as @patrickvonplaten indicated.

**Comment by StefanHeng** [comment]
Sure, do we keep the 2 branches (line 113), i.e. support a case where we don't filter out -100? 

I'm fine either way. I added padding as an option because creating that extra mask token would consume extra memory.
